### PR TITLE
Repository Review

### PR DESCRIPTION
**Pull request by avidant**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,59 @@
-# test-bayes
+# Test Bayes
+
+[![Build Status](https://travis-ci.org/Test-Bayes/test-bayes.svg?branch=master)](https://travis-ci.org/Test-Bayes/test-bayes/)
+[![Coverage Status](https://coveralls.io/repos/github/Test-Bayes/test-bayes/badge.svg?branch=master)](https://coveralls.io/github/Test-Bayes/test-bayes?branch=master)
```
Comment (avidant) [inline README.md]: Lose this. It isn't working and is broken anyways

File: README.md [hunk 2]
```diff
@@ -1 +1,59 @@
...
+Test Bayes runs on top of JUnit and requires the addition of a simple annotation on top of the testclasses you want reordered.
+
+Test Bayes uses a combination of Bayes' theorem, running probabilities, total probabilities, test duration, and conditional probability to reorder the tests.
+
+## Pre-Requisites
+The minimum requirements to develop Test Bayes is:
+ - Maven
+ - Java 8
+
+## Usage
```
Comment (avidant) [inline README.md]: Update with new instructions

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,6 @@
+.idea/
+.idea/*
+target/
+target/*
+*.iml
+*.class
```
Comment (avidant) [inline .gitignore]: add `log-data/`

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    <groupId>edu.uw.cse.testbayes</groupId>
+    <artifactId>testbayes</artifactId>
+    <version>1.0</version>
+    <packaging>jar</packaging>
+
+    <name>Test Bayes</name>
+    <description>A tool that reorders JUnit Tests so that tests fail fast and fail early.</description>
+    <url>http://www.github.com/Test-Bayes/test-bayes</url>
+
+    <!-- Figure out what license we want -->
```
Comment (avidant) [inline pom.xml]: Which license?

File: src/main/java/edu/uw/cse/testbayes/Main.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
+package edu.uw.cse.testbayes;
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/Main.java]: Lose this file

File: src/main/java/edu/uw/cse/testbayes/fileio/LogReader.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+/**
+ * Reads Logs from Test Runs and parses through the data
+ */
+public class LogReader extends TestLogIO {
+
+    // TODO: Make this a parameter of some sort
+    /**
+     * The number of files of logs to be read
+     */
+    private static final int RUNNING_AVERAGE = 15;
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/fileio/LogReader.java]: Make this a parameter

File: src/main/java/edu/uw/cse/testbayes/model/Bayes.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+     * @param s The name of the test
+     * @return total probability of test passing
+     */
+    public Probability getTestProb(String s) {
+        return new Probability(tots.get(s));
+    }
+
+    /**
+     * Returns the probability of s2 passing given s1 passed
+     * @param s1 The name of the test that already passed
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/model/Bayes.java]: Add line

File: src/main/java/edu/uw/cse/testbayes/model/Bayes.java [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                prob.addDenominator(1);
+                tot.put(test, prob);
+            }
+        }
+        return tot;
+    }
+
+    /**
+     * Uses the execution map provided to produce Conditional probability that a test will pass based on other tests passing
+     * @param testExecs A map with all test execution details
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/model/Bayes.java]: Blank Line

File: src/main/java/edu/uw/cse/testbayes/model/Probability.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        if(p1.denominator != p2.denominator) {
+            throw new IllegalStateException();
+        }
+
+        return p1.numerator - p2.numerator;
+    }
+
+    /**
+     * Returns the String representation of the Probability
+     * @return String representation of the Probability
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/model/Probability.java]: Blank line.

Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/model/Probability.java]: Add clarity of what type of representation

File: src/main/java/edu/uw/cse/testbayes/runner/IndividualClassRunner.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+            } catch (InvocationTargetException e) {
+                e.printStackTrace();
+                System.exit(1);
+            }
+        }
+    }
+
+    /**
+     * Sorts the methods in the methods list based on annotations
+     * @param methods List of methods to be sorted
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/runner/IndividualClassRunner.java]: Empty line

File: src/main/java/edu/uw/cse/testbayes/runner/IndividualClassRunner.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+                methods.remove(i);
+                i--;
+            } else {
+                nameToMethod.put(methods.get(i).toString(), methods.get(i));
+            }
+        }
+    }
+
+    // TODO: Add this documentation
+    /**
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/runner/IndividualClassRunner.java]: Missing documentation

File: src/main/java/edu/uw/cse/testbayes/runner/MethodInvocation.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
+package edu.uw.cse.testbayes.runner;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class MethodInvocation {
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/runner/MethodInvocation.java]: Missing docs

File: src/main/java/edu/uw/cse/testbayes/utils/FileNameUtils.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+import java.io.File;
+
+/**
+ * Provides static methods for file and directory names
+ */
+public class FileNameUtils {
+
+    /**
+     * Generates the name of the log file
+     * @return String with the name of the log file
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/utils/FileNameUtils.java]: Blank line

File: src/main/java/edu/uw/cse/testbayes/utils/FileNameUtils.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+     */
+    public static String createFileName() {
+        long timestamp = System.currentTimeMillis();
+        String filename = "log-data" + File.separator + timestamp + "-" + System.getProperty("user.name") + ".txt";
+        return filename;
+    }
+
+    /**
+     * Gets the name of the directory in which log files are stored
+     * @return String with the name of the directory
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/utils/FileNameUtils.java]: Blank Line

File: src/main/java/edu/uw/cse/testbayes/utils/LoggerUtils.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
+package edu.uw.cse.testbayes.utils;
+
+import org.apache.maven.plugin.logging.Log;
+
+public class LoggerUtils {
```
Comment (avidant) [inline src/main/java/edu/uw/cse/testbayes/utils/LoggerUtils.java]: Docs?

File: src/test/java/fileio/LogReaderTest.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+        map2 = new HashMap<>();
+        map2.put("public void package.Class.method1(String s) throws IOException", 100.23);
+        map2.put("public void package.Class.method2(String s) throws IOException", -100.23);
+
+        logData2 = new LogData(map2, false);
+    }
+
+    /**
+     * Tests that a file where the method names do not have any spaces is read correctly
+     * @throws FileNotFoundException
```
Comment (avidant) [inline src/test/java/fileio/LogReaderTest.java]: Blank Line. What's the error for?

File: src/test/java/fileio/LogWriterTest.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    @BeforeClass
+    public static void initialize() {
+        testname = "method";
+        duration = 100.23;
+        files = new HashSet<>();
+    }
+
+    /**
+     * Tests that a single test is written correctly
+     * @throws IOException in case of any IO issues
```
Comment (avidant) [inline src/test/java/fileio/LogWriterTest.java]: Blank Line

File: src/test/java/runner/BayesRunnerTest.java [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+import static org.junit.Assert.assertTrue;
+
+/**
+ * This class tests the Test Bayes individual class runner
+ */
+public class BayesRunnerTest {
+
+    /**
+     * Verify that individual class runner throws an exception for null class
+     * @throws InitializationError indicates runner class is null
```
Comment (avidant) [inline src/test/java/runner/BayesRunnerTest.java]: Blank line

File: src/test/java/runner/RandomRunnerTest.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.ArrayList;
+
+/**
+ * This class tests the Test Bayes individual class runner
+ */
+public class RandomRunnerTest {
+
+    /**
+     * Verify that individual class runner throws an exception for null class
+     * @throws InitializationError indicates runner class is null
```
Comment (avidant) [inline src/test/java/runner/RandomRunnerTest.java]: Blank line

File: src/test/java/runner/utilTestClasses/Test1.java [hunk 20]
```diff
@@ -0,0 +1,55 @@
+package runner.utilTestClasses;
+
+import edu.uw.cse.testbayes.runner.TestBayesIndividualClassRunner;
+import edu.uw.cse.testbayes.utils.LoggerUtils;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(TestBayesIndividualClassRunner.class)
```
Comment (avidant) [inline src/test/java/runner/utilTestClasses/Test1.java]: Documentation

File: src/test/java/runner/utilTestClasses/Test2.java [hunk 21]
```diff
@@ -0,0 +1,17 @@
+package runner.utilTestClasses;
+
+import org.junit.Ignore;
+import org.junit.Test;
+
+@Ignore
+public class Test2 {
```
Comment (avidant) [inline src/test/java/runner/utilTestClasses/Test2.java]: Documentation

File: src/test/java/runner/utilTestClasses/Test3.java [hunk 22]
```diff
@@ -0,0 +1,12 @@
+package runner.utilTestClasses;
+
+import org.junit.Ignore;
+import org.junit.Test;
+
+@Ignore
+public class Test3 {
```
Comment (avidant) [inline src/test/java/runner/utilTestClasses/Test3.java]: Docs

File: src/test/java/runner/utilTestClasses/Test4.java [hunk 23]
```diff
@@ -0,0 +1,27 @@
+package runner.utilTestClasses;
+
+import org.junit.Ignore;
+import org.junit.Test;
+
+@Ignore
+public class Test4 {
```
Comment (avidant) [inline src/test/java/runner/utilTestClasses/Test4.java]: Doc

File: src/test/java/runner/utilTestClasses/TestRunner.java [hunk 24]
```diff
@@ -0,0 +1,19 @@
+package runner.utilTestClasses;
+
+import edu.uw.cse.testbayes.runner.TestBayesIndividualClassRunner;
+import org.junit.Ignore;
+import org.junit.runner.RunWith;
+import org.junit.runners.Suite;
+
+@Ignore
```
Comment (avidant) [inline src/test/java/runner/utilTestClasses/TestRunner.java]: Doc

File: tests-file.txt [hunk 25]
```diff
@@ -0,0 +1,600 @@
+	@Test
```
Comment (avidant) [inline tests-file.txt]: Lose this file

**Comment by avidant** [comment]
Make epsilon a parameter

File: README.md (hunk 1 above)
Comment (AdityaJhamb) [reply README.md]: Done

File: .gitignore (hunk 3 above)
Comment (AdityaJhamb) [reply .gitignore]: Done

File: src/main/java/edu/uw/cse/testbayes/model/Probability.java (hunk 9 above)
Comment (AdityaJhamb) [reply src/main/java/edu/uw/cse/testbayes/model/Probability.java]: Resolved

File: src/main/java/edu/uw/cse/testbayes/model/Bayes.java (hunk 7 above)
Comment (AdityaJhamb) [reply src/main/java/edu/uw/cse/testbayes/model/Bayes.java]: Resolved

File: src/main/java/edu/uw/cse/testbayes/model/Bayes.java (hunk 8 above)
Comment (AdityaJhamb) [reply src/main/java/edu/uw/cse/testbayes/model/Bayes.java]: Resolved

File: src/main/java/edu/uw/cse/testbayes/model/Probability.java (hunk 9 above)
Comment (AdityaJhamb) [reply src/main/java/edu/uw/cse/testbayes/model/Probability.java]: Resolved

File: src/test/java/runner/BayesRunnerTest.java (hunk 18 above)
Comment (AdityaJhamb) [reply src/test/java/runner/BayesRunnerTest.java]: Done

File: src/test/java/runner/RandomRunnerTest.java (hunk 19 above)
Comment (AdityaJhamb) [reply src/test/java/runner/RandomRunnerTest.java]: Done

File: src/test/java/fileio/LogWriterTest.java (hunk 17 above)
Comment (AdityaJhamb) [reply src/test/java/fileio/LogWriterTest.java]: Done

File: src/main/java/edu/uw/cse/testbayes/runner/IndividualClassRunner.java (hunk 10 above)
Comment (AdityaJhamb) [reply src/main/java/edu/uw/cse/testbayes/runner/IndividualClassRunner.java]: Done

File: src/main/java/edu/uw/cse/testbayes/utils/FileNameUtils.java (hunk 13 above)
Comment (AdityaJhamb) [reply src/main/java/edu/uw/cse/testbayes/utils/FileNameUtils.java]: Done

File: src/main/java/edu/uw/cse/testbayes/utils/FileNameUtils.java (hunk 14 above)
Comment (AdityaJhamb) [reply src/main/java/edu/uw/cse/testbayes/utils/FileNameUtils.java]: Done

File: README.md (hunk 2 above)
Comment (avidant) [reply README.md]: Done

File: src/main/java/edu/uw/cse/testbayes/Main.java (hunk 5 above)
Comment (avidant) [reply src/main/java/edu/uw/cse/testbayes/Main.java]: Done

File: src/main/java/edu/uw/cse/testbayes/fileio/LogReader.java (hunk 6 above)
Comment (avidant) [reply src/main/java/edu/uw/cse/testbayes/fileio/LogReader.java]: Done

File: src/main/java/edu/uw/cse/testbayes/utils/LoggerUtils.java (hunk 15 above)
Comment (avidant) [reply src/main/java/edu/uw/cse/testbayes/utils/LoggerUtils.java]: Done

File: src/main/java/edu/uw/cse/testbayes/runner/IndividualClassRunner.java (hunk 11 above)
Comment (adavya) [reply src/main/java/edu/uw/cse/testbayes/runner/IndividualClassRunner.java]: done

File: src/test/java/runner/utilTestClasses/Test2.java (hunk 21 above)
Comment (adavya) [reply src/test/java/runner/utilTestClasses/Test2.java]: done

File: src/test/java/runner/utilTestClasses/Test3.java (hunk 22 above)
Comment (adavya) [reply src/test/java/runner/utilTestClasses/Test3.java]: done

File: src/test/java/runner/utilTestClasses/Test4.java (hunk 23 above)
Comment (adavya) [reply src/test/java/runner/utilTestClasses/Test4.java]: done

File: pom.xml (hunk 4 above)
Comment (AdityaJhamb) [reply pom.xml]: Resolved

**Pull request merged**